x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -504,27 +613,130 @@ public void messageReceived(DataNodeRequest request, TransportChannel channel, T
+                configuration,
+                request.aliasFilters(),
+                ActionListener.wrap(searchContexts -> {
+                    var computeContext = new ComputeContext(sessionId, searchContexts, configuration, null, exchangeSink);
+                    runCompute(parentTask, computeContext, request.plan(), ActionListener.wrap(driverProfiles -> {
+                        // don't return until all pages are fetched
+                        exchangeSink.addCompletionListener(
+                            ContextPreservingActionListener.wrapPreservingContext(
+                                ActionListener.releaseAfter(
+                                    listener.map(nullValue -> new DataNodeResponse(driverProfiles)),
+                                    () -> exchangeService.finishSinkHandler(sessionId, null)
+                                ),
+                                transportService.getThreadPool().getThreadContext()
+                            )
+                        );
+                    }, e -> {
+                        exchangeService.finishSinkHandler(sessionId, e);
+                        listener.onFailure(e);
+                    }));
                 }, e -> {
                     exchangeService.finishSinkHandler(sessionId, e);
                     listener.onFailure(e);
-                }));
-            }, e -> {
-                exchangeService.finishSinkHandler(sessionId, e);
-                listener.onFailure(e);
-            }));
+                })
+            );
+        }
+    }
+
+    public static final String CLUSTER_ACTION_NAME = EsqlQueryAction.NAME + "/cluster";
+
+    private class ClusterRequestHandler implements TransportRequestHandler<ClusterComputeRequest> {
+        @Override
+        public void messageReceived(ClusterComputeRequest request, TransportChannel channel, Task task) {
+            ChannelActionListener<ClusterComputeResponse> listener = new ChannelActionListener<>(channel);
+            if (request.plan() instanceof ExchangeSinkExec == false) {
+                listener.onFailure(new IllegalStateException("expected exchange sink for a remote compute; got " + request.plan()));
+                return;
+            }
+            runComputeOnRemoteCluster(
+                request.clusterAlias(),
+                request.sessionId(),
+                (CancellableTask) task,
+                request.configuration(),
+                (ExchangeSinkExec) request.plan(),
+                Set.of(request.indices()),
+                request.originalIndices(),
+                listener
+            );
+        }
+    }
+
+    /**
+     * Performs compute for a single remote cluster.
+     * <p>
+     * Currently, the coordinator on the remote cluster collects pages from data nodes in the remote cluster and provides them to the
+     * coordinator of the main cluster. We can use a single exchange buffer for this purpose. However, we use two buffers so that we
```

Comment:
> We can use a single exchange buffer for this purpose. However, we use two buffers so that we

I think these sentences together are a bit confusing. It *looks* to me like we just use a buffer on the way in and transfer to a buffer on the way out. We *could* perform partial reduction here, but we haven't wired it up yet. Though I'm probably missing something.

Reply:
I've reworded this in https://github.com/elastic/elasticsearch/pull/101640/commits/2c026b6b10ec61fb838fe3fd464be8c5037360ea.